File: python-impl/src/leetcode/find_diagonal_order.py
def find_diagonal_order(mat: list[list[int]]) -> list[int]:
    m = len(mat)
    n = len(mat[0])

    result, intermediate = [], []
    for d in range(m + n - 1):
        if d < n:
            r = 0
            c = d
        else:
            r = d - n + 1
            c = n - 1
        intermediate.clear()
        while r < m and c > -1:
            intermediate.append(mat[r][c])
            r += 1
            c -= 1
        if len(intermediate) > 1 and d % 2 == 0:
            intermediate.reverse()
        result.extend(intermediate)
    return result

File: golang-impl/src/leetcode/find_diagonal_order.go
package leetcode

import "sort"

func FindDiagonalOrder(mat [][]int) []int {
	m := len(mat)
	n := len(mat[0])
	var (
		res          []int
		intermediate []int
	)
	var r, c int
	for d := 0; d < m+n-1; d++ {
		if d < n {
			r = 0
			c = d
		} else {
			r = d - n + 1
			c = n - 1
		}
		intermediate = nil
		for r < m && c > -1 {
			intermediate = append(intermediate, mat[r][c])
			r++
			c--
		}
		if len(intermediate) > 1 && d%2 == 0 {
			sort.Sort(sort.Reverse(sort.IntSlice(intermediate)))
		}
		res = append(res, intermediate...)
	}
	return res
}
